pkg/config/validator.go

```diff
@@ -418,12 +555,108 @@ func (p *validator) validate(cfg map[string]string) (rt *Validated, ce *adapter.
 			continue
 		}
 		if re := p.validateServiceConfig(*ck, cfg[kk], true); re != nil {
-			return rt, ce.Appendf("serviceConfig", "failed validation").Extend(re)
+			rce = rce.Extend(re)
+		}
+	}
+
+	for _, kk := range keymap[constructors] {
+		ck := parseRulesKey(kk)
+		if ck == nil {
+			continue
+		}
+		if re := p.validateConstructorConfigs(cfg[kk]); re != nil {
+			rce = rce.Extend(re)
 		}
 	}
+
+	for _, kk := range keymap[actionRules] {
+		ck := parseRulesKey(kk)
+		if ck == nil {
+			continue
+		}
+		if re := p.validateRulesConfig(cfg[kk]); re != nil {
+			rce = rce.Extend(re)
+		}
+	}
+
+	if rce.Multi != nil && len(rce.Multi.Errors) != 0 {
+		// error has happened, we quit
+		return rt, ce.Appendf("Config", "failed validation").Extend(rce)
+	}
+
+	// everything is validated we can configure the handlers.
+	if re := p.buildHandlers(); re != nil {
+		return rt, ce.Appendf("Config", "failed validation").Extend(re)
+	}
+
 	return p.validated, nil
 }
 
+// TODO: What about partial failures. In the middle of building handlers,
+// adapter code might have made connections to the back-ends, should we
+// call close on built handlers in case there is an error after building few handlers.
+func (p *validator) buildHandlers() (ce *adapter.ConfigErrors) {
+
+	if err := p.setupHandler(p.actions, p.ctors, p.handlers, p.tmpls, p.typeChecker, p.descriptorFinder); err != nil {
+		return ce.Appendf("handlerConfig", "failed to configure handler: %v", err)
+	}
+
+	rce := &adapter.ConfigErrors{}
+	for handler, builder := range p.handlers {
+		if builder.isBroken {
+			// handler is broken, we should not build and cache it.
+			continue
+		}
+		if err := p.buildHandler(builder, handler); err != nil {
+			rce = rce.Extend(err)
+		}
+	}
+
+	if rce.Multi != nil && len(rce.Multi.Errors) != 0 {
```

Comment:
We go to great pains to ensure that all the methods on `adapter.ConfigErrors` keep the invariant that the value remains `nil` if extend/append are called with a `nil` arg (i.e. `ce.Extend(nil) == nil` if `ce == nil`). You shouldn't need to dig into the error, just a check against `nil` should be enough (`if rce != nil {  ... }`).